### PR TITLE
make topicalias mapping easier to understand

### PR DESCRIPTION
**Pull request by david-katz**

### CODE REVIEW

File: src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java [hunk 1]
```diff
@@ -18,53 +18,67 @@
 package org.mqttbee.mqtt.message.publish;
 
 import org.mqttbee.annotations.NotNull;
+import org.mqttbee.api.mqtt.datatypes.MqttTopic;
 import org.mqttbee.api.mqtt.mqtt5.message.publish.TopicAliasUsage;
-import org.mqttbee.mqtt.datatypes.MqttTopicImpl;
 
+import java.util.Collections;
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java]: unused import

File: src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java [hunk 2]
```diff
@@ -18,53 +18,67 @@
...
+
+        if (nextTopicAlias > topicAliasMaximum &&
+                topicAliasUsage == TopicAliasUsage.MAY_OVERWRITE) {
+            // used up all allowed topic aliases. Can only set a new one
+            // if allowed to overwrite an existing one.
+            topicAlias = random.nextInt(topicAliasMaximum) + 1;
+            // first remove the old mapping
+            hashMap.values().remove(topicAlias);
+            hashMap.put(topic.toString(), topicAlias);
+        } else {
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java]: This is not correct. In the following case we are assigning a topic alias which is greater than the topicAliasMaximum:
nextTopicAlias > topicAliasMaximum and topicAliasUsage != TopicAliasUsage.MAY_OVERWRITE

File: src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java [hunk 3]
```diff
@@ -18,53 +18,67 @@
...
-                if (topicAliasUsage == TopicAliasUsage.MAY_OVERWRITE) {
-                    topicAlias = random.nextInt(size) + 1;
-                    hashMap.put(topic.toString(), topicAlias);
-                }
-            } else {
-                topicAlias = nextTopicAlias;
-                hashMap.put(topic.toString(), topicAlias);
-                nextTopicAlias++;
-            }
+    public int set(@NotNull final MqttTopic topic, @NotNull final TopicAliasUsage topicAliasUsage) {
```
Comment (christianhoff) [inline src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java]: this is not really a set method. It is more a setIfAbsentAndGet. So I would probably split in setIfAbsent and then use the normal get method.

File: src/main/java/org/mqttbee/mqtt/handler/publish/MqttOutgoingQoSHandler.java [hunk 4]
```diff
@@ -155,13 +155,8 @@ private MqttPublishWrapper wrapPublish(
             isNewTopicAlias = false;
         } else {
             final MqttTopicImpl topic = publish.getTopic();
+            isNewTopicAlias = topicAliasMapping.createIfAbsent(topic, publish.getTopicAliasUsage());
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/handler/publish/MqttOutgoingQoSHandler.java]: What I do not like about this approach: we have to perform always 2 HashMap.get operations although not necessary

File: src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java [hunk 5]
```diff
@@ -18,53 +18,69 @@
...
-                hashMap.put(topic.toString(), topicAlias);
-                nextTopicAlias++;
-            }
+    public boolean createIfAbsent(@NotNull final MqttTopic topic, @NotNull final TopicAliasUsage topicAliasUsage) {
+        if (topicAliasUsage == TopicAliasUsage.MUST_NOT) {
+            return false;
+        }
+
+        // we have run out of topic aliases and cannot overwrite
+        if (nextTopicAlias > topicAliasMaximum && topicAliasUsage != TopicAliasUsage.MAY_OVERWRITE) {
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java]: I don't find this easier to understand

File: src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java [hunk 6]
```diff
@@ -18,53 +18,69 @@
...
+        if (topicAlias != DEFAULT_NO_TOPIC_ALIAS) {
+            return false;
+        }
+
+        // assign next available topic alias
+        if (nextTopicAlias <= topicAliasMaximum) {
+            topicAlias = nextTopicAlias;
+            nextTopicAlias++;
+        } else {
+            assert topicAliasUsage == TopicAliasUsage.MAY_OVERWRITE;
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java]: I also don't find this easier to understand

File: src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java (hunk 5 above)
Comment (david-katz) [reply src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java]: This is basically 1 to 1 the spec written as a logical expression and replaces a triple nested if-else.

**Comment by christianhoff** [comment]
Can be closed when PR 180 is merged.

**Pull request closed without merging**